src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs

```diff
@@ -304,46 +338,21 @@ private ConnectionCompleteParams GetConnectionCompleteParams(string connectionTy
                 connectionInfo.AddConnection(connectionParams.Type, connection);
 
                 // Add a cancellation token source so that the connection OpenAsync() can be cancelled
-                using (source = new CancellationTokenSource())
+                source = new CancellationTokenSource();
+                // Locking here to perform two operations as one atomic operation
+                lock (cancellationTokenSourceLock)
                 {
-                    // Locking here to perform two operations as one atomic operation
-                    lock (cancellationTokenSourceLock)
+                    // If the URI is currently connecting from a different request, cancel it before we try to connect
+                    CancellationTokenSource currentSource;
+                    if (cancelTupleToCancellationTokenSourceMap.TryGetValue(cancelKey, out currentSource))
```

Comment:
I'm pretty sure we could spruce up this code by replacing it with a AddOrUpdate call, but that might be going above and beyond for this code review.